### DISCUSSION TITLE
how to enable autosave mode?

### DISCUSSION THREAD

**Discussion by ousia** [orbitalquark/textadept#218] [Q&A] [answered]
I have just experienced the weirdest program crash since I started using Linux decades ago.

What I wrote in 20 minutes got lost (which is a collateral damage), but the main problem is that I didn’t got any OS warning.

I have no idea whether Textadept simply quit without asking before leaving unsaved buffers. There was also no segfault warning from the OS (_Fedora_).

All was apparently triggered by clicking twice on one of the lines from _[Message Buffer]*_ to which I copied the log data from a ConTeXt compilation ran in Geany. I intended to select and copy it. Then, Textadept vanished as if nothing happened. Of course, when I reopened it, all unsaved changes from the text file I was working on were gone.

To avoid this in the future, I would like to know whether I can create an autosave mode toggle for buffers. In part, this would be similar to the wrap mode toggle.

But first I need to know how to automatically save the current buffer each given amount of time.

Could someone help me here?

As a final remark: please, don’t get me wrong. I would love to offer the backtrace of the segfault or any similar info that helps to fix the bug. But before I can replicate the issue again and get the desired information, I need to prevent data loss.

**Comment by orbitalquark** [comment, marked as answer]
Oh no! I'm so sorry this happened to you :(

Textadept does not have an auto-save feature, but you can mimic one with something like this in your *~/.textadept/init.lua*:

  autosave = true -- global
  local interval = 60 -- seconds
  timeout(interval, function()
    if autosave and buffer.filename then buffer:save() end
    return true -- continue the timer indefinitely
  end)

Please note that if you have modules that utilize `events.BUFFER_BEFORE_SAVE` and `events.BUFFER_AFTER_SAVE` (e.g. linters), they will be run every 60 seconds and could fail (e.g. invalid syntax) and pop up the message buffer unexpectedly. If this happens to you, you could dry putting `ui.silent_print = true` before `buffer:save()`.

**Reply by ousia** [reply]
Many thanks for your reply, @orbitalquark.

How can I toggle activation/deactivation of your suggested function above?

Many thanks for your help.

**Reply by orbitalquark** [reply]
I've updated my reply with an `autosave` global that you can toggle on and off via the Tools > Command Entry. (`autosave = false` and `autosave = true`)

**Comment by ousia** [comment]
I tried to add this to my `init.lua` file:

```lua
--bindings['buffer:save_as()'] = nil

auto_save_enabled = false -- global

function auto_save_buffer()
  local interval = 5 -- seconds
  timeout(interval, function()
    if auto_save_enabled and buffer.filename and buffer.modify then buffer:save() end
    return true -- continue the timer indefinitely
  end)
end

keys['alt+s'] = function()
  if auto_save_enabled then
    auto_save_enabled = false
    auto_save_buffer()
  else
    auto_save_enabled = true
    auto_save_buffer()
  end
end
```

The first line tries to disable `ctrl+shift+s` for `buffer:save_as()`. I have no idea about how to do it.

I thought a function and key bindings would work here. But the code above doesn’t work at all 😅.

What am I doing egregiously wrong?

**Reply by rgieseke** [reply]
Not sure i fully understand what you're are trying to do, but 
```lua
keys['ctrl-S'] = nil
```
should disable that shortcut. (Or you could re-assign it to another function.)

**Comment by orbitalquark** [comment]
I haven't tried this, but you don't need the `auto_save_buffer()` function.

    keys['ctrl+shift+s'] = nil

    auto_save_enabled = false

    local interval = 5 -- seconds
    timeout(interval, function()
      if auto_save_enabled and buffer.filename and buffer.modify then buffer:save() end
      return true -- continue the timer indefinitely
    end)

    keys['alt+s'] = function() auto_save_enabled = not auto_save_enabled end

**Reply by ousia** [reply]
@orbitalquark,

only a minor question, how can I get auto-save only in the buffer in which I enabled it?

Many thanks for your help.

**Comment by orbitalquark** [comment]
Robert is right, `keys['ctrl+S']` should work. You don't need `shift+` like I suggested.

**Reply by rgieseke** [reply]
(I was unsure, so I did check https://github.com/orbitalquark/textadept/blob/default/modules/textadept/keys.lua)

**Comment by orbitalquark** [comment]
I would just use a `buffer.auto_save_enabled` field.

    local interval = 5 -- seconds
    timeout(interval, function()
      if buffer.auto_save_enabled and buffer.filename and buffer.modify then buffer:save() end
      return true -- continue the timer indefinitely
    end)

    keys['alt+s'] = function() buffer.auto_save_enabled = not buffer.auto_save_enabled end

**Reply by ousia** [reply]
Many thanks for your help, @orbitalquark.

**Comment by eric-hemasystems** [comment]
What about this as an alternative impl:

```lua
events.connect(events.UNFOCUS, io.save_all_files)
```

Advantages:

* I believe `buffer` just refers to the current buffer. Therefore in the other example implementations, if you made a change in a buffer and before the interval triggered switched to a different buffer it would not auto-save the previous buffer correct? And since you never know when that interval will happen that seems a bit shaky for a feature to give you confidence your stuff is on disk.
* It doesn't randomly write the file while you are in the middle of typing out something that might be syntactically invalid (possibly triggering various "watch" processes). Only when you change focus to another app (and presumably have finished your current text edit). Most of the time this would be saving it often enough although if you were in textadept for a long time without a focus switch you can always still `Ctrl-S` and explicitly save.

Any downside I'm not seeing?

**Reply by orbitalquark** [reply]
That sounds reasonable. The only downside I can think of is if you're in a "distraction free" kind of desktop mode and you don't leave Textadept for a long time, you won't see this benefit.

**Reply by eric-hemasystems** [reply]
Good point. I ended up also adding:

```lua
events.connect(events.BUFFER_BEFORE_SWITCH, io.save_all_files)
events.connect(events.VIEW_BEFORE_SWITCH, io.save_all_files)
```

This way even if you remain in Textadept but are just switching buffers or views it also saves as you switch your focus from one thing to another. Obviously if you are just editing a single buffer in a single view then either manually saving periodically or the timeout method works better for that workflow.
